src/faction_camp.cpp

```diff
@@ -709,60 +669,133 @@ void basecamp::get_available_missions( mission_data &mission_key, bool by_radio
-                                      "least one loot destination zone.\n"
-                                      "> Only items that are in the unsorted loot zone and not "
-                                      "in any other zone will be sorted.\n"
-                                      "Items that do not have a loot destination zone will be "
-                                      "sorted using the normal rules for automatic zone "
-                                      "sorting." ) );
-            mission_key.add( "Reset Sort Points", _( "Reset Sort Points" ), entry );
-            validate_assignees();
-            std::vector<npc_ptr> npc_list = get_npcs_assigned();
-            entry = string_format( _( "Notes:\n"
-                                      "Assign repeating job duties to NPCs stationed here.\n"
-                                      "Difficulty: N/A\n"
-                                      "Effects:\n"
-                                      "\n\nRisk: None\n"
-                                      "Time: Ongoing" ) );
-            mission_key.add( "Assign Jobs", _( "Assign Jobs" ), entry );
-        }
-    }
-
-    if( has_provides( "firewood" ) ) {
+        if( dir == base_dir ) {
+            // return legacy workers
+            comp_list npc_list = get_mission_workers( "_faction_upgrade_camp" );
+            if( !npc_list.empty() ) {
+                const base_camps::miss_data &miss_info = base_camps::miss_info[ "_faction_upgrade_camp" ];
+                entry = miss_info.action.translated();
+                bool avail = update_time_left( entry, npc_list );
+                mission_key.add_return( miss_info.ret_miss_id, miss_info.ret_desc.translated(),
+                                        cata::nullopt, entry, avail );
+            }
+            for( const basecamp_upgrade &upgrade : available_upgrades( dir ) ) {
+                gather_bldg = upgrade.bldg;
+                const base_camps::miss_data &miss_info = base_camps::miss_info[ "_faction_upgrade_camp" ];
+                comp_list npc_list = get_mission_workers( upgrade.bldg + "_faction_upgrade_camp" );
+                if( npc_list.empty() && !upgrade.in_progress ) {
+                    entry = om_upgrade_description( upgrade.bldg );
+                    mission_key.add_start( miss_info.miss_id + upgrade.bldg,
+                                           miss_info.desc + " " + upgrade.name, cata::nullopt, entry, upgrade.avail );
+                } else if( !npc_list.empty() && upgrade.in_progress ) {
+                    entry = miss_info.action.translated();
+                    bool avail = update_time_left( entry, npc_list );
+                    mission_key.add_return( miss_info.ret_miss_id + upgrade.bldg,
+                                            miss_info.ret_desc + " " + upgrade.name, cata::nullopt, entry, avail );
+                }
+            }
+
+            // this used to be optional, but now it isn't.  Keep it at the same scope
```

Comment:
You can go ahead and remove the brackets, the comment, and de-indent the following section.